### ISSUE TITLE
Nodes regularly disappear from K8S even though available in EC2

### ISSUE THREAD

**Issue by janroesner** [kubernetes/kops#10763]
**1. What `kops` version are you running? The command `kops version`, will display
 this information.**

Version 1.18.2

**2. What Kubernetes version are you running? `kubectl version` will print the
 version if a cluster is running or provide the Kubernetes version specified as
 a `kops` flag.**

Client Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.3", GitCommit:"1e11e4a2108024935ecfcb2912226cedeafd99df", GitTreeState:"clean", BuildDate:"2020-10-14T12:50:19Z", GoVersion:"go1.15.2", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.15", GitCommit:"73dd5c840662bb066a146d0871216333181f4b64", GitTreeState:"clean", BuildDate:"2021-01-13T13:14:05Z", GoVersion:"go1.13.15", Compiler:"gc", Platform:"linux/amd64"}

**3. What cloud provider are you using?**

AWS

**4. What commands did you run?  What is the simplest way to reproduce this issue?**

```
kops create cluster \
--name $KOPS_CLUSTER_NAME \
--cloud aws \
--master-size t2.small \
--master-zones=eu-central-1a \
--master-count=1 \
--node-size t2.small \
--zones=eu-central-1a,eu-central-1b,eu-central-1c \
--node-count=4 \
--kubernetes-version=1.18.15 \
--network-cidr=10.0.0.0/16 \
--networking=canal \
--authorization=RBAC \
--ssh-public-key="~/.ssh/id_rsa.pub" \
--topology=private
```

Afterwards I needed to update the cluster config as follows, to be able to run MetricServer:

```
kops edit cluster
```
I changed: 

```
    kubelet:
      anonymousAuth: false
```

to:

```
    kubelet:
      anonymousAuth: false
      authenticationTokenWebhook: true
      authorizationMode: Webhook
```

**5. What happened after the commands executed?**

The cluster initializes properly. After deployment of applications it runs without issues for 12 to 24 hours. At a certain point in time nodes disappear from the cluster. `kubectl get nodes` shows them not just "NotReady" but seems not to be able to retrieve any node related information. Lens shows 3 taints. Oftentimes a second node becomes stale a few minutes after the first one. The nodes __still are__ available in EC2 though and seem to run properly. The cluster can be stabelized only by killing the nodes in EC2 and have the AutoscalingGroup bring up new ones. This behaviour repeats again and again.

**6. What did you expect to happen?**

The nodes should not "disappear" from the cluster. 

**7. Please provide your cluster manifest. Execute
  `kops get --name my.example.com -o yaml` to display your cluster manifest.
  You may want to remove your cluster name and other sensitive information.**

```yaml
apiVersion: kops.k8s.io/v1alpha2
kind: Cluster
metadata:
  creationTimestamp: "2021-02-07T15:32:20Z"
  generation: 1
  name: cluster-five.xx-yy.zz.k8s.local
spec:
  api:
    loadBalancer:
      type: Public
  authorization:
    rbac: {}
  channel: stable
  cloudProvider: aws
  configBase: s3://kops-five-xx-yy/cluster-five.xx-yy.zz.k8s.local
  containerRuntime: docker
  etcdClusters:
  - cpuRequest: 200m
    etcdMembers:
    - instanceGroup: master-eu-central-1a
      name: a
    memoryRequest: 100Mi
    name: main
  - cpuRequest: 100m
    etcdMembers:
    - instanceGroup: master-eu-central-1a
      name: a
    memoryRequest: 100Mi
    name: events
  iam:
    allowContainerRegistry: true
    legacy: false
  kubelet:
    anonymousAuth: false
    authenticationTokenWebhook: true
    authorizationMode: Webhook
  kubernetesApiAccess:
  - 0.0.0.0/0
  kubernetesVersion: 1.18.15
  masterInternalName: api.internal.cluster-five.xx-yy.zz.k8s.local
  masterPublicName: api.cluster-five.xx-yy.zz.k8s.local
  networkCIDR: 10.0.0.0/16
  networking:
    canal: {}
  nonMasqueradeCIDR: 100.64.0.0/10
  sshAccess:
  - 0.0.0.0/0
  subnets:
  - cidr: 10.0.32.0/19
    name: eu-central-1a
    type: Private
    zone: eu-central-1a
  - cidr: 10.0.64.0/19
    name: eu-central-1b
    type: Private
    zone: eu-central-1b
  - cidr: 10.0.96.0/19
    name: eu-central-1c
    type: Private
    zone: eu-central-1c
  - cidr: 10.0.0.0/22
    name: utility-eu-central-1a
    type: Utility
    zone: eu-central-1a
  - cidr: 10.0.4.0/22
    name: utility-eu-central-1b
    type: Utility
    zone: eu-central-1b
  - cidr: 10.0.8.0/22
    name: utility-eu-central-1c
    type: Utility
    zone: eu-central-1c
  topology:
    dns:
      type: Public
    masters: private
    nodes: private

---

apiVersion: kops.k8s.io/v1alpha2
kind: InstanceGroup
metadata:
  creationTimestamp: "2021-02-07T15:32:20Z"
  labels:
    kops.k8s.io/cluster: cluster-five.xx-yy.zz.k8s.local
  name: master-eu-central-1a
spec:
  image: 099720109477/ubuntu/images/hvm-ssd/ubuntu-focal-20.04-amd64-server-20210119.1
  machineType: t2.small
  maxSize: 1
  minSize: 1
  nodeLabels:
    kops.k8s.io/instancegroup: master-eu-central-1a
  role: Master
  subnets:
  - eu-central-1a

---

apiVersion: kops.k8s.io/v1alpha2
kind: InstanceGroup
metadata:
  creationTimestamp: "2021-02-07T15:32:20Z"
  labels:
    kops.k8s.io/cluster: cluster-five.xx-yy.zz.k8s.local
  name: nodes
spec:
  image: 099720109477/ubuntu/images/hvm-ssd/ubuntu-focal-20.04-amd64-server-20210119.1
  machineType: t2.small
  maxSize: 4
  minSize: 4
  nodeLabels:
    kops.k8s.io/instancegroup: nodes
  role: Node
  subnets:
  - eu-central-1a
  - eu-central-1b
  - eu-central-1c
```

**8. Please run the commands with most verbose logging by adding the `-v 10` flag.
  Paste the logs into this report, or in a gist and provide the gist link here.**

-

**9. Anything else do we need to know?**
I set up 1.18.3, 1.18.9 and 1.18.15, all show the same behaviour. I ran exactly the same pod load for 2 1/2 years on 1.8.6 and for 10 month on 1.12.xx without ever having that issue. The problem is also unlikely to be related to AWS credits, since the load of all nodes is way below 20% all of the time and the credits are fully loaded. But when one (or two) of the nodes disappear, there's always a 40% to 50% spike in CPU load of nodes and master visible in CloudWatch. The CPU credits decrease, but not to a dramatic level.

**Comment by olemarkus** [comment]
I assume you roll _all_ nodes after doing that config change?

To triage further, you can look at the kubelet logs on the disappeared nodes. It should say why it cannot stay connected to the cluster.

**Comment by janroesner** [comment]
Hello @olemarkus,

I did not roll all nodes but created the cluster via:

```
kops update cluster --yes
```
I did NOT make that change to a cluster already running. Do I have to do a rolling update even before cluster initialization?

Regarding kubelet logs: 2 nodes just died few minutes ago again, but I can not connect to the pods via kubectl, therefore no access to the logs. Again, both instances were still available, normal and without issues in EC2.

**Comment by olemarkus** [comment]
In order to read the kubelet log you need to ssh into your ssh instance and run `journalctl -u kubelet`

**Comment by janroesner** [comment]
I will do that, next time an instance dies and come back. Thank you.

**Comment by olemarkus** [comment]
Add a bastion:
https://github.com/kubernetes/kops/blob/master/docs/bastion.md

**Comment by justinsb** [comment]
There's also some behaviour in kube-controller-manager which should remove NotReady nodes by shutting them down in EC2, though it has a failsafe where if a majority of nodes are unavailable it assumes that the control plane is at fault and doesn't take action.  (It may be a majority of nodes in one AZ, I can't recall exactly) 

If you're able to SSH into the node as @olemarkus suggests, we may get a good clue there as to what is happening.

Also, if you're able to do `kubectl get nodes <broken-node-name> -oyaml` or `kubectl describe node <broken-node-name>` when it is NotReady, there may be an indication there as to why the node is NotReady (e.g. if disk is full, I _think_ that gets surfaced also).  `kubectl describe` is probably better because it will also print the events.

**Comment by janroesner** [comment]
I close this issue, since my cluster is running stable for 24 hours now. Should the problems re-occur, I will open a new issue.

For everyone who might have similar issues, there's just one thing, that I changed in my pod configs: resource requests and resource limits. Beforehand I configured requests and limits both for CPU and mem very generously. Especially the limits exceeded the capacity by far. That was no problem, since, my processes usually consume very little resources. Obviously sth in the K8s behavior changed since 1.12 which led to a situation I was not able to determine in detail and which ended in hosts going missing in the cluster. With updated resource requests and limits close to the real values the cluster seems to be very stable.
